src/core/layout/qgslayoutitemchart.cpp

```diff
@@ -0,0 +1,390 @@
+  mGathererTimer.setSingleShot( true );
+  connect( &mGathererTimer, &QTimer::timeout, this, &QgsLayoutItemChart::gatherData );
+}
+
+int QgsLayoutItemChart::type() const
+{
+  return QgsLayoutItemRegistry::LayoutChart;
+}
+
+QIcon QgsLayoutItemChart::icon() const
+{
+  return QgsApplication::getThemeIcon( QStringLiteral( "/mLayoutItemChart.svg" ) );
+}
+
+QgsLayoutItemChart *QgsLayoutItemChart::create( QgsLayout *layout )
+{
+  return new QgsLayoutItemChart( layout );
+}
+
+void QgsLayoutItemChart::setPlot( QgsPlot *plot )
+{
+  Qgs2DPlot *plot2d = dynamic_cast<Qgs2DPlot *>( plot );
+  if ( !plot2d )
+  {
+    delete plot;
+    return;
+  }
+
+  // Logic to minimise plot data refresh to bare minimum
+  bool requireRefresh = !mPlot || !plot;
+  if ( mPlot && plot )
+  {
+    if ( mPlot->type() != plot->type() )
+    {
+      requireRefresh = true;
+    }
+    else
+    {
+      Qgs2DXyPlot *oldPlot2dXy = dynamic_cast<Qgs2DXyPlot *>( mPlot.get() );
+      Qgs2DXyPlot *newPlot2dXy = dynamic_cast<Qgs2DXyPlot *>( plot2d );
+      if ( oldPlot2dXy && oldPlot2dXy->xAxis().type() != newPlot2dXy->xAxis().type() )
+      {
+        requireRefresh = true;
+      }
+    }
+  }
+
+  mPlot.reset( plot2d );
+  if ( requireRefresh )
+  {
+    refresh();
+  }
+
+  emit changed();
+}
+
+void QgsLayoutItemChart::setSourceLayer( QgsVectorLayer *layer )
+{
+  if ( layer == mVectorLayer.get() )
+  {
+    return;
+  }
+
+  mVectorLayer.setLayer( layer );
+  refresh();
+
+  emit changed();
+}
+
+
+void QgsLayoutItemChart::setSortFeatures( bool sorted )
+{
+  if ( mSortFeatures == sorted )
+  {
+    return;
+  }
+
+  mSortFeatures = sorted;
+  refresh();
+
+  emit changed();
+}
+
+void QgsLayoutItemChart::setSortAscending( bool ascending )
+{
+  if ( mSortAscending == ascending )
+  {
+    return;
+  }
+
+  mSortAscending = ascending;
+  refresh();
+
+  emit changed();
+}
+
+void QgsLayoutItemChart::setSortExpression( const QString &expression )
+{
+  if ( mSortExpression == expression )
+  {
+    return;
+  }
+
+  mSortExpression = expression;
+  refresh();
+
+  emit changed();
+}
+
+void QgsLayoutItemChart::setSeriesList( const QList<QgsLayoutItemChart::SeriesDetails> &seriesList )
+{
+  if ( mSeriesList == seriesList )
+  {
+    return;
+  }
+
+  mSeriesList = seriesList;
+  refresh();
+
+  emit changed();
+}
+
+void QgsLayoutItemChart::draw( QgsLayoutItemRenderContext & )
+{
+}
+
+void QgsLayoutItemChart::paint( QPainter *painter, const QStyleOptionGraphicsItem *itemStyle, QWidget * )
+{
+  if ( !mLayout || !painter || !painter->device() )
+  {
+    return;
+  }
+
+  if ( !shouldDrawItem() )
+  {
+    return;
+  }
+
+  if ( !mPlot )
+    return;
+
+  QPaintDevice *paintDevice = painter->device();
+  if ( !paintDevice )
+    return;
+
+  QRectF thisPaintRect = rect();
+  if ( qgsDoubleNear( thisPaintRect.width(), 0.0 ) || qgsDoubleNear( thisPaintRect.height(), 0 ) )
+    return;
+
+  if ( mLayout->renderContext().isPreviewRender() )
```

Comment:
I think we should detect invalid plot settings here (eg no series set), and render a text label "missing chart data" or similar